### ISSUE TITLE
Bug: ROLLBACK cannot be parsed

### ISSUE THREAD

**Issue by GitMensch** [mridoni/gix#67]
```sql
              EXEC SQL AT :DBS
                 COMMIT
              END-EXEC.
              EXEC SQL AT :DBS
                 ROLLBACK
              END-EXEC.
```

The first is accepted, the second one results in

> error: syntax error, unexpected END_EXEC, expecting SELECT or TOKEN or UPDATE or FOR

Reason: the scanner explicit requests a "." behind that, adding this lets the code compile

**Comment by GitMensch** [comment]
... and when the syntax is used so a compilation works:

```cobol
GIXSQL*    EXEC SQL AT :DBS
GIXSQL*       ROLLBACK.
GIXSQL*    END-EXEC.
``` 
breaks at runtime as it copies the "." in the statement

> "SQ0003": "ROLLBACK .\000"

With the result

> SQLCODE=-0000010007|SQLSTATE=42617->ERROR:  syntax error at or near "."
> LINE 1: ROLLBACK .


It _seems_ to work on both sides if I change https://github.com/mridoni/gix/blob/8dd2cdbeaacf72c9300bccfe0c6ef23dfc09160d/libgixpp/gix_esql_scanner.ll#L517
```diff
- 	"ROLLBACK"[ ]+"." { 
+ 	"ROLLBACK { 
```

**Comment by mridoni** [comment]
Already fixed and tested in my internal repository. 

I am thinking of releasing v1.0.11 in the beginning of next week (including at least fixes for #66, #67, #73) then work on #47, #50, #65, #68, that are more complex and need more testing. I am thinking of scheduling those for a v1.0.12 release.

**Comment by GitMensch** [comment]
Issue planning is also be good to be done in a separate issue and/or via milestones ;-)
A bugfix release "sooner" sounds good, I've lost a bit the track about the issues so I can't tell how helpful this would be, it will be in any case.

The test environment is running on our side, currently bails out with a 
~~~~
DB-CODE = 10007
DB-MSG  = ERROR:  DECLARE CURSOR can only be used in transaction blocks
~~~~

It worked with ocesql but I _think_ there were some parts about transactions - will need to check with someone knowing the COBOL code better on Monday.

**Comment by mridoni** [comment]
That error is straight from PostgreSQL: you cannot DECLARE a cursor outside a transaction (this is why many of my test cases have a START TRANSACTION at the beginning, just after a connection is made. The alternative would be to do just like in the other drivers (can't remember if MySQL,ODBC or both) and "emulate" the cursor, issuing just the SELECT when opening and keeping the resultset handle around. But since libpq and PostgreSQL have native cursor support, I thought it was better to handle it this way.

**Comment by GitMensch** [comment]
Sounds reasonable so far. I'll check the COBOL codebase and the way it was handled with multiple other precompilers, including 2 others for PostgreSQL on Monday and report then.
This issue is just about the parsing part (actually lexing) which you have fixed so can be closed once checked in.

**Comment by mridoni** [comment]
The ROLLBACK fix generated a small regression (`ROLLBACK TO SAVEPOINT` is handled in the parser, and it shouldn't be), but nothing that can't be fixed before v1.0.11.

**Comment by GitMensch** [comment]
> The alternative would be to do just like in the other drivers (can't remember if MySQL,ODBC or both) and "emulate" the cursor, issuing just the SELECT when opening and keeping the resultset handle around.

I haven't checked in detail how the (really broken) citpgsql did it, but I've got the information that it explicit required a `BEGIN` before using cursors, too.
ocesql does _not_ need it, so it may be reasonable to have an internal driver attribute (and/or environment option) for compatibility that is passed and handled in the pgqsql driver. Its implementation is in https://github.com/opensourcecobol/Open-COBOL-ESQL/blob/44ba99e752880e4ab75d089143e86d7901f194f5/dblib/ocpgsql.c#L105-L135 and dblib/ocesql.c (4 places with `"BEGIN"`)  that would have to be moved to the driver like https://github.com/opensourcecobol/Open-COBOL-ESQL/blob/44ba99e752880e4ab75d089143e86d7901f194f5/dblib/ocesql.c#L696-L705 doing an auto-transaction - or is that already implemented?

**Comment by mridoni** [comment]
I have to check the code (I don't remember the details) but there is some of that going on (inserting `BEGIN` or similar statements), mainly to support autocommit, and this is why `START/BEGIN TRANSACTION`, `COMMIT` and `ROLLBACK` are "trapped" by the library. I will look into it but I am going to track this in a separate issue (BTW: `ROLLBACK` is now parsed correctly, including `ROLLBACK TO SAVEPOINT`, implementation will be in v1.0.11).

**Comment by GitMensch** [comment]
This issue is solved with 1.0.11, the ocesql compat has a separate issue already --> closed.
